### PR TITLE
Query ties

### PR DESCRIPTION
**Pull request by greatbacon**
WIP draft for querying & rendering entities & their ties

### CODE REVIEW

File: src/lib/vocab/tie/tieRepo.ts [hunk 1]
```diff
@@ -18,4 +18,36 @@ export const tieRepo = (db: Database) =>
...
+			WITH RECURSIVE paths (source_id, dest_id, type, created, path) AS (
+        	SELECT t.source_id, t.dest_id, t.type, t.created, ARRAY[t.source_id, t.dest_id]
+        	FROM ties t WHERE source_id=${directory_id}
+    		UNION ALL
+        	SELECT t.source_id, t.dest_id, t.type,t.created, p.path || ARRAY[t.dest_id]
+        	FROM paths p
+        	JOIN ties t
+        	ON p.dest_id = t.source_id AND t.dest_id != ALL(p.path)
+				)
+			SELECT * FROM paths;
```
Comment (ryanatkn) [inline src/lib/vocab/tie/tieRepo.ts:48]: I think this is what we talked about:

```suggestion
			SELECT source_id, dest_id, type, created FROM paths;
```

File: src/lib/vocab/tie/tieRepo.ts [hunk 2]
```diff
@@ -18,4 +18,36 @@ export const tieRepo = (db: Database) =>
...
+			FROM ties t 
+			JOIN (SELECT entity_id FROM entities WHERE space_id=${space_id}) as e 
+			ON e.entity_id = t.source_id OR e.entity_id = t.dest_id;
+			`;
+			console.log('[db] space ties', ties);
+			return {ok: true, value: ties};
+		},
+		//TODO BLOCK add test that checks this
+		//TODO BLOCK add migration unique constraint on ties & indexes
+		navigateTies: async (directory_id: number): Promise<Result<{value: Tie[]}>> => {
```
Comment (ryanatkn) [inline src/lib/vocab/tie/tieRepo.ts:36]: Did we say `filterBySource` or `filterBySourceId` or something? And then `directory_id` becomes `source_id`. And we'll write similarly named queries for the dest and bidirectional versions.

File: src/lib/ui/view/Todo.svelte [hunk 3]
```diff
@@ -21,6 +23,38 @@
...
+	$: tiesResult?.then((data) => {
+		if (data.ok) {
+			ties = data.value.ties;
+		}
+	});
+
+	$: itemsByEntity = $entities && ties ? toItemsByEntity($entities, ties) : null;
+
+	let entityById: Map<number, Readable<Entity>> | null = null;
+	$: entityById = $entities && new Map($entities.map((e) => [get(e).entity_id, e]));
```
Comment (ryanatkn) [inline src/lib/ui/view/Todo.svelte:38]: I think one of our next big pair programming tasks may be how to query entities better than this. Looks like a fun challenge :robot:

File: src/lib/ui/TodoItem.svelte [hunk 4]
```diff
@@ -39,54 +50,109 @@
...
-	<div class="markup formatted">
-		<div class="signature">
-			<Avatar name={toName($persona)} icon={toIcon($persona)} showIcon={false} />
-			{#if $entity.updated}
-				updated {format(new Date($entity.updated), 'Pp')}
-			{:else}
-				created {format(new Date($entity.created), 'Pp')}
+{#if renderEntity($entity)}
+	<li
+		style="--hue: {hue}"
```
Comment (ryanatkn) [inline src/lib/ui/TodoItem.svelte:77]: new syntax dropped:

```suggestion
		style:--hue={hue}
```

File: src/lib/db/migrations/00015-index-ties.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  `;
+
+	await sql`
+    CREATE INDEX idx_dest_id
+    ON ties (dest_id);
+  `;
+
+	await sql`
+    ALTER TABLE ties
+    ADD CONSTRAINT UC_ties UNIQUE (source_id,dest_id,type);
```
Comment (ryanatkn) [inline src/lib/db/migrations/00015-index-ties.js:16]: v cool

File: src/lib/ui/TodoItem.svelte (hunk 4 above)
Comment (greatbacon) [reply src/lib/ui/TodoItem.svelte]: I tried this but it threw an error about a type mismatch (number can't be String) so I'm reverting it.

**Review by ryanatkn — APPROVED** [review]
awesome :100:

File: src/lib/ui/view/Todo.svelte [hunk 6]
```diff
@@ -21,6 +23,39 @@
...
+		const map: Map<Readable<Entity>, Array<Readable<Entity>>> = new Map();
+		for (const tie of ties) {
+			if (tie.type !== 'HasItem') continue;
+			const source = entities.find((e) => get(e).entity_id === tie.source_id)!;
+			const dest = entities.find((e) => get(e).entity_id === tie.dest_id)!;
+			let items = map.get(source);
+			if (!items) {
+				map.set(source, (items = []));
+			}
+			items.push(dest);
```
Comment (ryanatkn) [inline src/lib/ui/view/Todo.svelte:55]: whoops I sidetracked us in conversation, this is the point I dropped:

```suggestion
			if (!items.includes(dest)) items.push(dest);
```

**Pull request merged**